778 Update conditional start event on parent change

smbea opened the pull request:
Related to https://github.com/bpmn-io/bpmn-js/issues/778

pinussilvestrus commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,59 @@
+        if (def.get('camunda:variableEvents')) {
+
+          // if shape within event subpprocess changes parent
+          if (is(oldParent, 'bpmn:SubProcess') && !is(newParent, 'bpmn:SubProcess') && oldParentBusinessObject.get('triggeredByEvent')) {
```
The order of the conditions could be confusing (oldParent, newParent, oldParent...). Maybe we can do something like

```js
 if (
  is(oldParent, 'bpmn:SubProcess') && 
  oldParentBusinessObject.get('triggeredByEvent') && 
  !is(newParent, 'bpmn:SubProcess')
) { 
  ... 
}
 ```

pinussilvestrus reviewed (commented):
I like the descriptive implementation and test cases, it was easy to understand what's going on - good job 👍  Some things to consider codewise (cf. comments).

I saw that the commit message (cf. https://github.com/camunda/camunda-bpmn-moddle/pull/77/commits/4a2b45a9e28a32ee2b6c7e50734b5618e0706d4d) is quite extensive - maybe due to some refactorings & updates.

Let's keep it clean 

```sh
fix(moddle): remove variableEvents when StartEvent parent changes 

Related to bpmn-io/bpmn-js#778
```

smbea commented:
Changes overview:

@barmac I updated the way the objects were being modified, as we had talked about (using `bpmnFactory` and `moddleCopy` to create and modify a copy of each object)

@pinussilvestrus I fixed the consistency discrepancies and updated the test diagram to include more than one `eventDefinition` in the conditional `startEvent`

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+      });
+    }
+
+    modeling.updateProperties(shape, { 'eventDefinitions': definitionsCopy });
```
We want to update event definitions only when it's necessary. As it is here in the code, the `eventDefinitions` property will be updated whenever one moves any shape as there are no conditions applied.

barmac replied:
Also, let's not update the properties if no `camunda:variableEvents` had to be replaced.

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+
+      definitionsCopy.forEach(function(def,i) {
+
+        var newDef = bpmnFactory.create('bpmn:ConditionalEventDefinition');
+        moddleCopy.copyElement(def, newDef);
+        newDef.$parent = def.$parent;
```
We make a copy for each event definition which seems to be unnecessary when `camunda:variableEvents` is not found. Also, there are more types of event definitions, e.g. `bpmn:TimerEventDefinition`. We don't want to look into them or replace them.

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+
+          // if shape within an event subpprocess changes parent
+          if (
+            is(oldParent, 'bpmn:SubProcess') &&
```
If a user already has a broken diagram, i.e. where a process child has `camunda:variableEvents`, we want to fix that. Thus, there is no need for the old parent check.

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+
+  injector.invoke(CommandInterceptor, this);
+
+  this.postExecuted(['shape.move'], function(context) {
```
We should also react to `shape.create` to fix the issue where possible.

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+        if (def.get('camunda:variableEvents')) {
+
+          // if shape within an event subpprocess changes parent
+          if (
+            is(oldParent, 'bpmn:SubProcess') &&
+            oldParentBusinessObject.get('triggeredByEvent') &&
+            !is(newParent, 'bpmn:SubProcess')
+          ) {
```
This condition is checked multiple times while it could be checked once and earlier in the function.

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+    var shapeBusinessObject = getBusinessObject(shape);
+    var eventDefinitions, definitionsCopy;
+
+    // if shape is a startEvent
```
```suggestion
```
Duplicate of the code.

barmac commented on lib/RemoveVariableEventBehaviour.js:
```diff
@@ -0,0 +1,72 @@
+        // if event definitions have variableEvents property
+        if (def.get('camunda:variableEvents')) {
+
+          // if shape within an event subpprocess changes parent
```
We want to also fix the incorrect state. It's not necessary for the shape to change a parent. If it has `camunda:variableEvents` and it's incorrect, we should remove the property.

barmac reviewed (changes requested):
I've left a couple of comments. Overall, the solution is heading into right direction, but some additional work needs to be done.

smbea commented:
Changes summary:
- The conditions for the property removal were changed according to the comments made
- Copy of eventDefinitions object and parent check is now only done once
- Flag was added to check if there was a need to update the eventDefinitions, instead of always updating
- Missing test case was added

barmac commented:
I've renamed a couple of variables, changed formatting, and squashed the commits. Ready to merge :)
